quaddtype/numpy_quaddtype/_quaddtype_main.pyi

```diff
@@ -0,0 +1,134 @@
+from typing import Any, Literal, TypeAlias, final
+
+import numpy as np
+from typing_extensions import Never, Self, override
+
+_Backend: TypeAlias = Literal["sleef", "longdouble"]
+_IntoQuad: TypeAlias = QuadPrecision | float | str
+_CastsQuad: TypeAlias = _IntoQuad | np.floating[Any] | np.integer[Any] | np.bool_
+
+@final
+class QuadPrecDType(np.dtype[QuadPrecision]):  # type: ignore[misc, type-var]  # pyright: ignore[reportGeneralTypeIssues, reportInvalidTypeArguments]
+    def __new__(cls, /, backend: _Backend = "sleef") -> Self: ...
+
+    # `numpy.dtype` overrides
+    names: None  # pyright: ignore[reportIncompatibleVariableOverride]
+    @property
+    @override
+    def alignment(self) -> Literal[16]: ...
+    @property
+    @override
+    def itemsize(self) -> Literal[16]: ...
+    @property
+    @override
+    def name(self) -> Literal["QuadPrecDType128"]: ...
+    @property
+    @override
+    def byteorder(self) -> Literal["|"]: ...
+    @property
+    @override
+    def char(self) -> Literal["\x00"]: ...  # type: ignore[override]  # pyright: ignore[reportIncompatibleMethodOverride]
+    @property
+    @override
+    def kind(self) -> Literal["\x00"]: ...  # type: ignore[override]  # pyright: ignore[reportIncompatibleMethodOverride]
+    @property
+    @override
+    def num(self) -> Literal[-1]: ...  # type: ignore[override]  # pyright: ignore[reportIncompatibleMethodOverride]
+    @property
+    @override
+    def shape(self) -> tuple[()]: ...
+    @property
+    @override
+    def ndim(self) -> Literal[0]: ...
+    @property
+    @override
+    def fields(self) -> None: ...
+    @property
+    @override
+    def base(self) -> Self: ...
+    @property
+    @override
+    def subdtype(self) -> None: ...
+    @property
+    @override
+    def hasobject(self) -> Literal[False]: ...
+    @property
+    @override
+    def isbuiltin(self) -> Literal[0]: ...
+    @property
+    @override
+    def isnative(self) -> Literal[True]: ...
+    @property
+    @override
+    def isalignedstruct(self) -> Literal[False]: ...
+    @override
+    def __getitem__(self, key: Never, /) -> Self: ...  # type: ignore[override]
+
+@final
+class QuadPrecision:  # NOTE: It doesn't inherit from `np.generic` which is type-unsafe
```

Comment:
To elaborate a bit on this comment:
For all other dtypes (except `StringDType`, which doesn't even have a scalar type), the `dtype.type` is some subclass of `np.generic`. That means that it *should* be safe to do e.g. `arr.dtype.type.dtype`, since `np.generic.dtype`. But for `QuadPrecDType().dtype` there exists no `.type`.
This is called *type-unsafe* because there are situation where type-checkers are not able to detect these cases that will cause a `AttributeError` to be raised at runtime. 
This is of course not limited to `.dtype`, and there are (many) more examples that could go wrong like this.

Reply:
In an ongoing work we subclassing the scalar to `PyFloatingArrType_Type` i.e. dtype.type can be a subclass of `np.generic`

```
In [4]: issubclass(npq.QuadPrecDType().type, np.generic)
Out[4]: True
```

This is currently waiting for one PR https://github.com/numpy/numpy/pull/29763 to get merged in NumPy and then we add this change here as well

Reply:
does this resolve the type-unsafety?

Reply:
I take it that `PyFloatingArrType_Type` means that it'll subclass `np.floating` then?
Either way, it'll indeed be resolved then :)

Reply:
When it comes to that, and you need any help updating the stubs, I'd be happy to :)

Reply:
> I've really no idea how you generated them :)

By writing them myself haha